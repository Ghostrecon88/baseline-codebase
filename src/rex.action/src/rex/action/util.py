"""

    rex.action.util
    ===============

    :copyright: 2017, Prometheus Research, LLC

"""

__all__ = ('get_action_key',)

def get_action_key(parent_id, action_local_id):
    key = '%s@%s' % (parent_id, action_local_id)
    return key
