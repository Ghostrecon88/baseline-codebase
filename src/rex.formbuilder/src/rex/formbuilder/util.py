

import os
import fcntl
import tempfile
import shutil

class FileLock(object):

    def __init__(self, filename, create=False):
        self.filename = filename
        self.descriptor = None
        if create:
            if not os.path.isfile(self.filename):
                open(self.filename, 'a').close()
        else:
            if not os.path.isfile(self.filename):
                raise IOError("File %s does not exist" % self.filename)

    def __enter__(self):
        self.descriptor = open(self.filename, 'r')
        fcntl.flock(self.descriptor, fcntl.LOCK_EX)

    def __exit__(self, *args, **kwds):
        fcntl.flock(self.descriptor, fcntl.LOCK_UN)
        self.descriptor.close()
        self.descriptor = None


