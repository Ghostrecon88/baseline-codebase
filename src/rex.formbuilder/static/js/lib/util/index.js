/*
 * Copyright (c) 2015, Prometheus Research, LLC
 */

'use strict';

var Ajax = require('./Ajax');
var format = require('./format');
var isEmpty = require('./isEmpty');
var isEmptyLocalization = require('./isEmptyLocalization');
var merge = require('./merge');


module.exports = {
  Ajax,
  format,
  isEmpty,
  isEmptyLocalization,
  merge
};

