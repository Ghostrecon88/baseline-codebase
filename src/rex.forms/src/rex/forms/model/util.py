

import os
import fcntl
import tempfile
import shutil


def savefile(filename, data):
    f = tempfile.NamedTemporaryFile(delete=False)
    f.write(data)
    f.close()
    try:
        shutil.move(f.name, filename)
    finally:
        os.unlink(f.name)


class FileLock(object):

    def __init__(self, filename, create=False):
        self.filename = filename
        self.descriptor = None
        if create and not os.path.isfile(self.filename):
            open(self.filename, 'w').close()

    def __enter__(self):
        self.descriptor = open(self.filename, 'r')
        fcntl.flock(self.descriptor, fcntl.LOCK_EX)

    def __exit__(self, *args, **kwds):
        fcntl.flock(self.descriptor, fcntl.LOCK_UN)
        self.descriptor.close()
        self.descriptor = None


