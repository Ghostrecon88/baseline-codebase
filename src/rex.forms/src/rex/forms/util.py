#
# Copyright (c) 2016, Prometheus Research, LLC
#


from rex.instrument import Assessment, Subject


__all__ = (
    'preview_calculation_results',
)


def preview_calculation_results(
        instrument_version,
        calculation_set,
        data,
        assessment=None):
    if not calculation_set:
        return {}

    # Validate the Assessment Data
    assessment_impl = Assessment.get_implementation()
    assessment_impl.validate_data(
        data,
        instrument_definition=instrument_version.definition,
    )

    # Make some temporary objects so we can execute the calcs.
    if not assessment:
        subject = Subject.get_implementation()('fake')
        assessment = assessment_impl(
            'fake',
            subject,
            instrument_version,
            data,
            status=assessment_impl.STATUS_COMPLETE,
        )
    else:
        assessment.data = data
        assessment.status = assessment_impl.STATUS_COMPLETE

    # Execute the calculations
    return calculation_set.execute(assessment=assessment)

