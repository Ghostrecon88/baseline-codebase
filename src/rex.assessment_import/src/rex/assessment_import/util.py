import csv
from collections import OrderedDict

from rex.core import Error
from rex.instrument.util import get_implementation
from .interface import Instrument, Assessment


def get_assessment_templates(instrument_version,
                             additional_template_fields=None):
    additional_template_fields = additional_template_fields or OrderedDict()
    instrument = Instrument.create(instrument_version,
                                   additional_template_fields)
    return instrument.templates


def import_assessment_data(instrument_version,
                           obj_name,
                           assessment_data,
                           additional_template_fields=None,
                           additional_assessment_data=None):
    additional_template_fields = additional_template_fields or OrderedDict()
    instrument = Instrument.create(instrument_version,
                                   additional_template_fields)
    template = instrument.templates.get(obj_name)
    if not template:
        return
    try:
        reader = csv.DictReader(assessment_data)
    except Exception, exc:
        raise Error("Unable to read csv.", exc)
    for row in reader:
        try:
            assessment = Assessment.save(instrument,
                                         obj_name,
                                         row,
                                         additional_assessment_data)
        except Error, exc:
            print exc
            continue
