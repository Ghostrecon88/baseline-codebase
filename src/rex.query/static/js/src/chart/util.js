/**
 * @flow
 */

import type {QueryPipeline, QueryAtom, Context} from '../model/types';
import type {Chart} from './model';

import {inferTypeAtPath, aggregate} from '../model/Query';
import {editor} from '../model/QueryOperation';
import {getNavigation} from '../model/QueryNavigation';
import {getUsedAttributes} from './model';

export function getColumnOptions(
  context: Context,
): Array<{label: string, value: string}> {
  const nav = getNavigation(context);
  return Array.from(nav.values()).map(nav => ({
    label: nav.label,
    value: nav.value,
  }));
}

export function getQuery(
  query: QueryPipeline,
  data: any,
): {query: ?QueryPipeline, data: any} {
  if (query.pipeline.length === 1) {
    return {query: null, data};
  } else {
    if (query.pipeline[1].name === 'define') {
      if (data != null) {
        data = data[Object.keys(data)[0]];
      }
      return {query: query.pipeline[1].binding.query, data};
    } else {
      return {query: null, data};
    }
  }
}

export function enrichQuery(query: QueryPipeline, chart: Chart): QueryPipeline {
  const focus = getQuery(query, null).query;
  if (focus == null) {
    return query;
  }
  let e = editor(query, focus);
  for (const attrName of getUsedAttributes(chart)) {
    let editAtCompletion;
    const type = inferTypeAtPath(focus.context.prev, [attrName]);
    if (type.card === 'seq' && type.name === 'record' && type.entity != null) {
      editAtCompletion = ensurePipelineHasCount;
    }
    e = e.growNavigation({path: [attrName], editAtCompletion});
  }
  return e.getQuery();
}

function ensurePipelineHasCount(pipe: QueryAtom[]): QueryAtom[] {
  const last = pipe[pipe.length - 1];
  if (last != null && last.name === 'aggregate') {
    return pipe;
  } else {
    return pipe.concat(aggregate('count'));
  }
}
