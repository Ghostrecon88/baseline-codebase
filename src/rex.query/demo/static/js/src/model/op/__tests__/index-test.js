import * as jestMatchers from 'jest-matchers';
import * as matchers from 'jest-matchers/build/matchers';

import {def, pipeline, navigate, select, filter} from '../../Query';
import * as qp from '../../QueryPointer';
import {stripContext} from '../../__tests__/util';
import {insertAfter, removeAt} from '../index';

const study = navigate('study');

let pointerPath = pointer =>
  pointer
    ? pointer.path.map(item => item.join('.')).join(':')
    : null;

describe('insertAfter()', function() {

  it('individual!name', function() {
    let query = navigate('individual');
    let pointer = qp.make(query);
    let {query: nextQuery} = insertAfter({pointer, selected: null}, navigate('name'));
    expect(nextQuery).toMatchSnapshot();
  });

  it('[individual]!name', function() {
    let query = pipeline(navigate('individual'));
    let pointer = qp.make(query);
    let {query: nextQuery} = insertAfter({pointer, selected: null}, navigate('name'));
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual.sample!name', function() {
    let query = pipeline(navigate('individual'), navigate('sample'));
    let pointer = qp.make(query);
    let {query: nextQuery} = insertAfter({pointer, selected: null}, navigate('name'));
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual!sample.name', function() {
    let query = pipeline(navigate('individual'), navigate('name'));
    let pointer = qp.select(qp.make(query), ['pipeline', 0]);
    let {query: nextQuery} = insertAfter({pointer, selected: null}, navigate('sample'));
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual:define(a = sample!name)', function() {
    let query = pipeline(
      navigate('individual'),
      def('a', navigate('sample'))
    );
    let pointer = qp.select(
      qp.make(query),
      ['pipeline', 1],
      ['binding', 'query']
    );
    let {query: nextQuery} = insertAfter({pointer, selected: null}, navigate('name'));
    expect(nextQuery).toMatchSnapshot();
  });

});

describe('removeAt()', function() {

  it('individual.name!', function() {
    let query = pipeline(navigate('individual'), navigate('name'));
    let pointer = qp.select(qp.make(query), ['pipeline', 1]);
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual!.name', function() {
    let query = pipeline(navigate('individual'), navigate('name'));
    let pointer = qp.select(qp.make(query), ['pipeline', 0]);
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual:select(a = sample.name!)', function() {
    let query = pipeline(
      navigate('individual'),
      select({
        a: pipeline(navigate('sample'), navigate('name'))
      })
    );
    let pointer = qp.select(
      qp.make(query),
      ['pipeline', 1],
      ['select', 'a']
    );
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual:select(a = [sample!])', function() {
    let query = pipeline(
      navigate('individual'),
      select({
        a: pipeline(navigate('sample')),
      })
    );
    let pointer = qp.select(
      qp.make(query),
      ['pipeline', 1],
      ['select', 'a'],
      ['pipeline', 0]
    );
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual:select(a = sample.name!, b = sample)', function() {
    let query = pipeline(
      navigate('individual'),
      select({
        a: pipeline(navigate('sample'), navigate('name')),
        b: navigate('sample'),
      })
    );
    let pointer = qp.select(
      qp.make(query),
      ['pipeline', 1],
      ['select', 'a']
    );
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

  it('individual:define(a := sample!)', function() {
    let query = pipeline(
      navigate('individual'),
      def('a', pipeline(navigate('sample'))),
    );
    let pointer = qp.select(
      qp.make(query),
      ['pipeline', 1],
      ['binding', 'query']
    );
    let {query: nextQuery} = removeAt({pointer, selected: null});
    expect(nextQuery).toMatchSnapshot();
  });

});

