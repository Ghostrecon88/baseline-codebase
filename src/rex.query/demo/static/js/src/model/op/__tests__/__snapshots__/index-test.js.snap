exports[`insertAfter() [individual]!name 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`insertAfter() individual!name 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`insertAfter() individual!sample.name 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "sample",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`insertAfter() individual.sample!name 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "sample",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`insertAfter() individual:define(a = sample!name) 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "binding": Object {
        "name": "a",
        "query": Object {
          "context": Object {
            "domain": <DOMAIN>,
            "prev": "<PREV CONTEXT>",
            "scope": "<SCOPE [empty]>",
            "type": null,
          },
          "name": "pipeline",
          "pipeline": Array [
            Object {
              "context": Object {
                "domain": <DOMAIN>,
                "prev": "<PREV CONTEXT>",
                "scope": "<SCOPE [empty]>",
                "type": null,
              },
              "name": "navigate",
              "path": "sample",
            },
            Object {
              "context": Object {
                "domain": <DOMAIN>,
                "prev": "<PREV CONTEXT>",
                "scope": "<SCOPE [empty]>",
                "type": null,
              },
              "name": "navigate",
              "path": "name",
            },
          ],
        },
      },
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "define",
    },
  ],
}
`;

exports[`removeAt() individual!.name 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "name",
    },
  ],
}
`;

exports[`removeAt() individual.name! 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
  ],
}
`;

exports[`removeAt() individual:define(a := sample!) 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
  ],
}
`;

exports[`removeAt() individual:select(a = [sample!]) 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
  ],
}
`;

exports[`removeAt() individual:select(a = sample.name!) 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
  ],
}
`;

exports[`removeAt() individual:select(a = sample.name!, b = sample) 1`] = `
Object {
  "context": Object {
    "domain": <DOMAIN>,
    "prev": "<PREV CONTEXT>",
    "scope": "<SCOPE [empty]>",
    "type": null,
  },
  "name": "pipeline",
  "pipeline": Array [
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "here",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "navigate",
      "path": "individual",
    },
    Object {
      "context": Object {
        "domain": <DOMAIN>,
        "prev": "<PREV CONTEXT>",
        "scope": "<SCOPE [empty]>",
        "type": null,
      },
      "name": "select",
      "select": Object {
        "b": Object {
          "context": Object {
            "domain": <DOMAIN>,
            "prev": "<PREV CONTEXT>",
            "scope": "<SCOPE [empty]>",
            "type": null,
          },
          "name": "navigate",
          "path": "sample",
        },
      },
    },
  ],
}
`;
